File: tetrahedron.js
class GeometryLibrary {

}

class Point {
    constructor(x, y, z) {
        this.x = x
        this.y = y
        this.z = z

    }

    equals(other) {
        return this.x === other.x && this.y === other.y && this.z === other.z
    }   
}

class Tetrahedron extends GeometryLibrary {
    constructor(p1, p2, p3, p4) {
        super()
        this.p1 = p1
        this.p2 = p2
        this.p3 = p3
        this.p4 = p4
    }

    equals(other) {
        return (this.p1.equals(other.p1) && this.p2.equals(other.p2) && this.p3.equals(other.p3) && this.p4.equals(other.p4))
    }

    distance(p1, p2) {
        return Math.sqrt((p1.x - p2.x) ** 2 + (p1.y - p2.y) ** 2 + (p1.z - p2.z) ** 2)
    }

    // function for the centroid here
    // check center of mass for a tetrahedron
    //  equation for the center of mass for a solid tetrahedron = 1/4 * (v1 + v2 + v3 + v4)
    // v1 to v4 being vertices 
    //  link for the calculation -> https://math.stackexchange.com/questions/1592128/finding-center-of-mass-for-tetrahedron
    centroid() {
        const x = (this.p1.x + this.p2.x + this.p3.x + this.p4.x) / 4
        const y = (this.p2.y + this.p2.y + this.p3.y + this.p4.y) / 4
        const z = (this.p1.z + this.p2.y + this.p3.y + this.p4.z) / 4
        return new Point(x, y, z)
    }


    // surface area of a tetrahedron
    surfaceArea() {
        const a = this.distance(this.p1, this.p2)
        const b = this.distance(this.p1, this.p3)
        const c = this.distance(this.p1, this.p4)
        const d = this.distance(this.p2, this.p3)
        const e = this.distance(this.p2, this.p4)
        const f = this.distance(this.p3, this.p4)

        const s1 = (a + d + e) / 2
        const s2 = (a + b + f) / 2
        const s3 = (b + d + c) / 2
        const s4 = (c + e + f) / 2

        const area1 = Math.sqrt(s1 * (s1 - a) * (s1 - d) * (s1 - e))
        const area2 = Math.sqrt(s2 * (s2 - a) * (s2 - b) * (s2 - f))
        const area3 = Math.sqrt(s3 * (s3 - b) * (s3 - d) * (s3 - c))
        const area4 = Math.sqrt(s4 * (s4 - c) * (s4 - e) * (s4 - f))

        return area1 + area2 + area3 + area4

    }
}


// applying the class while creating a new instance
const p1 = new Point(0, 0, 0)
const p2 = new Point(1, 0, 0)
const p3 = new Point(0, 1, 0)
const p4 = new Point(0, 0, 1)

// all point classes are params here
// this makes the instance of the tetrahedron complete
const tetrahedron = new Tetrahedron(p1, p2, p3, p4)

// centroid use here
const centroid = tetrahedron.centroid()
console.log(`Centroid: ${centroid.x}, ${centroid.y}, ${centroid.z}`)


const surfaceArea = tetrahedron.surfaceArea()

console.log(`Surface Area: ${surfaceArea}`)

File: tetrahedron.cs
using System;
using System.Linq;

namespace GeometryLibrary
{
    public class Point3D
    {
        public double X { get; set; }
        public double Y { get; set; }
        public double Z { get; set; }

        public Point3D(double x, double y, double z)
        {
            X = x;
            Y = y;
            Z = z;
        }
    }

    public class Tetrahedron
    {
        private Point3D[] vertices;
        
        public Tetrahedron(Point3D v1,Point3D v2, Point3D v3, Point3D v4)
        {
            vertices = new Point3D[] { v1, v2, v3, v4};

        }

        public static bool operator == (Tetrahedron t1, Tetrahedron t2)
        {
            // the logic to compare the two tetrahedra here
            if (ReferenceEquals(t1, null))
            {
                return ReferenceEquals(t2, null);
            }

            if (ReferenceEquals(t2, null))
            {
                return false;
            }

        // Assuming two tetrahedra are equal if they have the same vertices in any order
        return t1.vertices.All(v1 => t2.vertices.Any(v2 => v1.Equals(v2))) &&
               t2.vertices.All(v1 => t1.vertices.Any(v2 => v1.Equals(v2)));
        }

        public static bool operator !=(Tetrahedron t1, Tetrahedron t2)
        {
            return !(t1 == t2);
        }

        public Point3D Centroid()
        {
            double x = (vertices[0].X + vertices[1].X + vertices[2].X + vertices[3].X ) / 4;
            double y = (vertices[0].Y + vertices[1].Y + vertices[2].Y + vertices[3].Y ) / 4;
            double z = (vertices[0].Z + vertices[1].Z + vertices[2].Z + vertices[3].Z ) / 4;
            return new Point3D(x, y, z);
        }

        public double Area(Point3D p1, Point3D p2, Point3D p3)
        {
            double a = Distance(p1, p2);
            double b = Distance(p1, p3);
            double c = Distance(p2, p3);

            double s = (a + b + c) / 2.0;
            
            return Math.Sqrt(s * (s - a) * (s - b) * (s - c));
        }

        private double Distance(Point3D p1, Point3D p2)
        {
            return Math.Sqrt(Math.Pow(p1.X - p2.X , 2) + Math.Pow(p1.Y - p2.Y, 2) + Math.Pow(p1.Z - p2.Z, 2));
        }

        public double SurfaceArea()
        {
            double area1 = Area(vertices[0], vertices[1], vertices[2]);
            double area2 = Area(vertices[0], vertices[1], vertices[3]);
            double area3 = Area(vertices[0], vertices[2], vertices[3]);
            double area4 = Area(vertices[1], vertices[2], vertices[3]);

            return area1 + area2 + area3 + area4;
        }

        public override bool Equals(object obj)
        {
            if(obj == null || GetType() != obj.GetType())
            {
                return false;
            }

            return this == (Tetrahedron)obj;
        }

        public override int GetHashCode()
        {
            return vertices[0].GetHashCode() ^ vertices[1].GetHashCode() ^ vertices[2].GetHashCode() ^ vertices[3].GetHashCode();
        }
    }
}

class Program
{
    static void Main(string[] args)
    {
        GeometryLibrary.Point3D p1 = new GeometryLibrary.Point3D(0, 0, 0);
        GeometryLibrary.Point3D p2 = new GeometryLibrary.Point3D(1, 0, 0);
        GeometryLibrary.Point3D p3 = new GeometryLibrary.Point3D(0, 1, 0);
        GeometryLibrary.Point3D p4 = new GeometryLibrary.Point3D(0, 0, 1);

        GeometryLibrary.Tetrahedron tetra = new GeometryLibrary.Tetrahedron(p1, p2, p3, p4);

        GeometryLibrary.Point3D centroid = tetra.Centroid();
        double surfaceArea = tetra.SurfaceArea();

        Console.WriteLine($"Centroid: ({centroid.X}, {centroid.Y}, {centroid.Z})");
        Console.WriteLine($"Surface Area: {surfaceArea}");

    }
}